File: 229-Majority Element II-求众数 II/229-Majority Element II-求众数 II.js
/**
 * @param {number[]} nums
 * @return {number[]}
 * 摩尔投票法：摩尔投票法的核心思想为对拼消耗。首先我们考虑最基本的摩* * 尔投票问题，比如找出一组数字序列中出现次数大于总数1/2​的数字（并且假 * 设这个数字一定存在）。
 * 可以直接推出：大于1/n（n为数组长度）的元素最多存在n - 1个元素。
 *
 */
// var majorityElement = function(nums) {
//     let map = new Map()
//     let res = []
//     let len = nums.length
//     let max = Math.floor(len / 3)
//     for(let i = 0; i < len; i++){
//         if(map.has(nums[i])){
//             map.set(nums[i], map.get(nums[i]) + 1)
//         }else{
//             map.set(nums[i], 1)
//         }
//     }
//     for(let k of map.keys()){
//         if(map.get(k) > max){
//             res.push(k)
//         }
//     }
//     return res
// };

var majorityElement = function(nums) {
    let element1 = 0;
    let element2 = 0;
    let vote1 = 0;
    let vote2 = 0;

    for (const num of nums) {
        if (vote1 > 0 && num === element1) { //如果该元素为第一个元素，则计数加1
            vote1++;
        } else if (vote2 > 0 && num === element2) { //如果该元素为第二个元素，则计数加1
            vote2++;
        } else if (vote1 === 0) { // 选择第一个元素
            element1 = num;
            vote1++;
        } else if (vote2 === 0) { // 选择第二个元素
            element2 = num;
            vote2++;
        } else { //如果三个元素均不相同，则相互抵消1次
            vote1--;
            vote2--;
        }
    }

    //统计筛选出的元素的数字
    let cnt1 = 0;
    let cnt2 = 0;
    for (const num of nums) {
        if (vote1 > 0 && num === element1) {
            cnt1++;
        }
        if (vote2 > 0 && num === element2) {
            cnt2++;
        }
    }
    // 检测元素出现的次数是否满足要求
    const ans = [];
    if (vote1 > 0 && cnt1 > Math.floor(nums.length / 3)) {
        ans.push(element1);
    }
    if (vote2 > 0 && cnt2 > Math.floor(nums.length / 3)) {
        ans.push(element2);
    }

    return ans;
};

File: 229-Majority Element II-求众数 II/229-Majority Element II-求众数 II.java
class Solution {
    // public List<Integer> majorityElement(int[] nums) {
    //     HashMap<Integer, Integer> map = new HashMap<Integer, Integer>();
    // 	HashSet<Integer> set = new HashSet<Integer>();
    //     int len = nums.length;
    //     int max = len/3;
    //     for(int i = 0; i < len; i++) {
    //     	if(map.containsKey(nums[i])) {
    //     		map.put(nums[i], map.get(nums[i]) + 1);
    //     	}else {
    //     		map.put(nums[i], 1);
    //     	}
    //     }
    //     for(int key : map.keySet()) {
    //     	if(map.get(key) > max) {
    //     		set.add(key);
    //     	}
    //     }
    //     return new ArrayList<Integer>(set);
    // }

    public List<Integer> majorityElement(int[] nums) {
        int v1 = 0;
    	int v2 = 0;
    	int cnt1 = 0;
    	int cnt2 = 0;
    	int len =nums.length;
    	HashSet<Integer> set = new HashSet<Integer>();
    	for(int i = 0; i < len; i++) {
    		if(nums[i] == v1 && cnt1 > 0) {
    			cnt1++;
    		}else if(nums[i] == v2 & cnt2 > 0) {
    			cnt2++;
    		}else if(cnt1 == 0) {
    			v1 = nums[i];
    			cnt1++;
    		}else if(cnt2 == 0) {
    			v2 = nums[i];
    			cnt2++;
    		}else  {
    			cnt1--;
    			cnt2--;
    		}
    	}
    	
    	int realC1 = 0;
    	int realC2 = 0;
    	for(int i = 0; i < len; i++) {
    		if(v1 == nums[i]) {
    			realC1++;
    		}else if(v2 == nums[i]) {
    			realC2++;
    		}
    	}
    	if(realC1 > len / 3) {
    		set.add(v1);
    	}
    	if(realC2 > len / 3) {
    		set.add(v2);
    	}
    	return new ArrayList<Integer>(set);
    }
}
